### PR TITLE
[NPU][Diffusion] MiniMax-H3 RainFusion end_step tail fallback

### PR DESCRIPTION
**Pull request by HAAZZZEEEE**
## Purpose

Add two knobs to the RAINFUSION_ATTN block-sparse path for MiniMax-H3 on
Ascend NPU:

1. **`end_step` tail fallback** — keep the last N denoise steps dense so the
   tail of the video is never sparsified. Together with `start_step`, the
   config now supports `{start_step, end_step, sparsity}` to express a
   dense → sparse → dense window across the denoise schedule.

2. **`precision` knob for the rf_v3 sparse-attention kernel** — select the
   execution precision per `block_sparse` config:
   - `mix` (default): EagleQBSA mixed precision (Q/K per-block INT8 + V
     per-channel FP8)
   - `fp8`: BSA FP8 path (Hadamard rotation + full FP8 block quantization of
     Q/K/V)
   - `bf16`: no quantization, pure BF16 sparse attention (previous behavior)

`BlockSparseSpec.precision` is validated against `RainFusionPrecision` and
forwarded via `backend_kwargs → RainFusionConfig → mindiesd.sparse_attention`.

## Test Plan

**vLLM Version:** 0.28.0+empty

**vLLM-Omni Commit:** this PR (846ad825 + 6d7b000c)

- Hardware: Ascend 950PR (A5) node, 4x NPU, 128 GB HBM per card
- Software: vllm-ascend latest main + MindIE-SD dev (rf_v3 EagleQBSA)
- Serve: MiniMax-H3 FL2VA, `RAINFUSION_ATTN`, layerwise offload,
  sparsity=0.8
- Request (standard recipe): T2VA, 1344x768, 50 steps, 24 fps,
  duration 8.7 s
- Verification probes (gated by `EQBSA_PROBE=1`):
  - `rainfusion_attn.py::_resolve_plan` prints per-step
    `dense(start)|sparse|dense(end)` decisions
  - `sparse_flash_attn_rf_v3.py::rain_fusion_attention_v3` prints
    `precision/sparsity/seqlen` before the `eagle_quant_block_sparse_attention`
    call

### A/B performance

| config | fallback window | E2E time | output |
|--------|----------------|----------|--------|
| dense (FLASH_ATTN) | — | ~561 s | 9.7 MB |
| bf16, no fallback | start=0 / end=0 | ~272 s | 9.7 MB |
| bf16 | start=8 / end=12 | ~390 s | 9.9 MB |
| mix (EagleQBSA), no fallback | start=0 / end=0 | ~241 s | 9.6 MB |
| mix (EagleQBSA) | start=8 / end=12 | ~375 s | 9.8 MB |

Component-wise breakdown:
- Sparse vs dense (FLASH_ATTN): **~51% faster** (561 s -> 272 s, bf16 no
  fallback)
- EagleQBSA quantization vs bf16 at the same no-fallback window: **~11%
  faster** (272 s -> 241 s)
- 8/12 fallback window protects quality at ~130 s cost vs no fallback
  (mix: 241 s -> 375 s; bf16: 272 s -> 390 s)
- Best practice (mix 8/12) vs dense baseline: **~33% faster**
  (561 s -> 375 s)
- Maximum gain (mix, no fallback) vs dense baseline: **~57% faster**
  (561 s -> 241 s)

Kernel-level confirmation:
- `[EQBSA-PROBE]` fired with `precision='mix'`, sparsity 0.8 in the sparse
  window (5,800 invocations for 8/12; all 50 steps for no-fallback) — the
  EagleQBSA kernel (`eagle_quant_block_sparse_attention`, Q/K per-block
  INT8 + V per-channel FP8) is actually dispatched on NPU, not silently
  falling back to dense
- `[PLAN-PROBE]` confirmed the dense -> sparse -> dense step pattern for
  the 8/12 window: steps 0-7 dense(start), 8-36 sparse, 37-48 dense(end)

### Video output comparison

[//]: # (Upload videos here)

- **A: dense (FLASH_ATTN)** — 

https://github.com/user-attachments/assets/d80047dd-5ba9-4e6c-b29c-a531bba724f1

- **B: bf16, no fallback** —

https://github.com/user-attachments/assets/e808d88f-6192-4b16-8548-8c4364680bcd

- **C: bf16, 8/12** —

https://github.com/user-attachments/assets/4a747a68-e7f6-4b33-9b6c-c141e4bf62d9

- **D: mix (EagleQBSA), no fallback** — 

Uploading out_768p_E_mix_nofb.mp4…

- **E: mix (EagleQBSA), 8/12** — [upload: video file]

https://github.com/user-attachments/assets/97740b0e-87db-4f3c-b294-4fb08eb1b8af

### CODE REVIEW

File: vllm_omni/diffusion/models/minimax_h3/minimax_h3_transformer.py [hunk 1]
```diff
@@ -43,6 +43,136 @@
 
     from vllm_omni.diffusion.data import OmniDiffusionConfig
 
+
+@torch.no_grad()
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/minimax_h3/minimax_h3_transformer.py]: do not change model file for a hardware attn backend

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py [hunk 2]
```diff
@@ -232,8 +236,17 @@ def _resolve_plan(self, attn_metadata: AttentionMetadata | None) -> RainFusionPl
...
         if is_forward_context_available():
             step_idx = get_forward_context().denoise_step_idx
+            total_steps = get_forward_context().total_denoise_steps
             if step_idx is not None and step_idx < rf.start_step:
                 return None
+            # Tail fallback: keep the last ``end_step`` denoise steps dense.
+            if (
+                rf.end_step > 0
+                and step_idx is not None
+                and total_steps is not None
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/rainfusion_attn.py:283]: MiniMax-H3 never publishes total_denoise_steps: its loop only sets the step and timestep, so this condition is always false and end_step has no effect. Please plumb num_steps through that producer and cover the first tail step with a regression test.

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py [hunk 3]
```diff
@@ -339,6 +352,7 @@ def _forward_sparse_npu(
             latent_shape_q=plan.latent_shape,
             latent_shape_k=plan.latent_shape,
             sparsity=self.rainfusion.sparsity,
+            precision=self.rainfusion.precision,
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/rainfusion_attn.py]: Older MindIE-SD releases accept this through **kwargs but silently ignore it, so selecting mix or fp8 can still run the old BF16 path. Please gate this on a MindIE-SD capability/version, or pin and document the required revision.

File: vllm_omni/__init__.py [hunk 4]
```diff
@@ -18,6 +18,17 @@
 # throw in patch if the versions differ.
 from .version import __version__, __version_tuple__  # isort:skip # noqa: F401
 
+# Early-register MindIE-SD custom ops (QBSA etc.) before CANN snapshots the
+# custom-op registry at the first regInfo lookup (vllm-ascend model
+# load/warmup). Without this, lazily-imported mindiesd ops (e.g.
+# eagle_quant_block_sparse_attention via RAINFUSION_ATTN) fail with EZ1014
+# "inferShape function does not exist" at first call. Mirrors PR #5891.
+try:
+    import mindiesd  # noqa: F401
```
Comment (david6666666) [inline vllm_omni/__init__.py]: This imports an NPU-only package for every vllm_omni import, including CUDA processes. If MindIE-SD is present, its vendor/custom-op side effects now run before platform selection. Could we keep the early registration behind an NPU-only bootstrap so CUDA startup stays isolated?

File: vllm_omni/diffusion/data.py [hunk 5]
```diff
@@ -1771,6 +1795,8 @@ def backend_kwargs(self) -> dict[str, Any] | None:
             bs = self.block_sparse
             kw["sparsity"] = bs.sparsity
             kw["start_step"] = bs.start_step
+            kw["end_step"] = bs.end_step
+            kw["precision"] = bs.precision
```
Comment (david6666666) [inline vllm_omni/diffusion/data.py:1804]: The existing block-sparse config tests still assert the old kwargs dictionaries, so this head will fail those tests. Please update them and add validation/forwarding coverage for both new knobs.

File: vllm_omni/diffusion/data.py [hunk 6]
```diff
@@ -1678,17 +1692,27 @@ class BlockSparseSpec:
...
     skip_layers: str | list[int] | None = None
     skip_layer_indices: set[int] | None = field(default=None, repr=False)
 
     def __post_init__(self) -> None:
         self.sparsity = _in_range(self.sparsity, "block_sparse.sparsity", 0.0, 1.0) or 0.0
         if self.start_step < 0:
             raise ValueError(f"block_sparse.start_step must be >= 0; got {self.start_step!r}.")
+        if self.end_step < 0:
+            raise ValueError(f"block_sparse.end_step must be >= 0; got {self.end_step!r}.")
+        if self.precision not in {p.value for p in RainFusionPrecision}:
```
Comment (david6666666) [inline vllm_omni/diffusion/data.py:1716]: RainFusionPrecision members pass this check because they subclass str, but RainFusionConfig later calls str(value), producing RainFusionPrecision.MIX instead of mix. Please normalize enum inputs to .value here (or use StrEnum).

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py (hunk 2 above)
Comment (HAAZZZEEEE) [reply vllm_omni/diffusion/attention/backends/rainfusion_attn.py:283]: Fixed. You're right — we plumbed num_steps through: the publish function now takes a total_steps argument and writes it via the new set_forward_context_denoise_total_steps setter; the H3 denoise loop passes num_steps, and the batched publish site in the pipeline passes len(extra[_STEP_SIGMAS_VIDEO]) - 1; regression tests cover the first tail step (step_idx == total_steps - end_step), the step just before the tail window (still sparse), and end_step=0 (never falls back) — three cases.

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py (hunk 3 above)
Comment (HAAZZZEEEE) [reply vllm_omni/diffusion/attention/backends/rainfusion_attn.py]: Gated on MindIE-SD capability. _mindiesd_supports_precision() inspects the installed sparse_attention signature for a precision parameter (cached once per process). When a non-bf16 precision is requested but the installed mindiesd lacks it, _forward_sparse_npu now raises RuntimeError instead of silently running the BF16 path. The requirement is also documented in docs/user_guide/diffusion/attention_backends/rainfusion.md.

File: vllm_omni/__init__.py (hunk 4 above)
Comment (HAAZZZEEEE) [reply vllm_omni/__init__.py]: Plz take a look, does this approach work? 

Kept the early registration but made it NPU-only. 

vllm_omni/__init__.py now wraps `import mindiesd` with hasattr(torch, "npu") and torch.npu.is_available() — CUDA processes short-circuit on importing vllm_omni and never touch the NPU-only package; NPU processes keep the pre-snapshot registration to prevent EZ1014. 

We intentionally kept the package-level location: an earlier attempt to move it to the NPU platform hook (PR #5891 style) was measured to run too late in spawn workers (after the CANN custom-op registry snapshot), which would regress EZ1014. The platform.py hook is kept as a harmless no-op backstop.

File: vllm_omni/diffusion/data.py (hunk 6 above)
Comment (HAAZZZEEEE) [reply vllm_omni/diffusion/data.py:1716]: Switched RainFusionPrecision from (str, Enum) to StrEnum, so str() returns the member value directly (RainFusionPrecision.MIX -> "mix") and the value survives the backend_kwargs() -> RainFusionConfig.from_backend_kwargs() round trip. Added test_block_sparse_precision_enum_input_normalized as a regression guard.

File: vllm_omni/diffusion/data.py (hunk 5 above)
Comment (HAAZZZEEEE) [reply vllm_omni/diffusion/data.py:1804]: Updated. 

The two existing assertions now expect the new keys: test_block_sparse_defaults_applied_when_backend_selected and test_block_sparse_skip_layers_selector_expanded include end_step/precision. test_block_sparse_invalid_values adds end_step=-1 and precision="bogus" for validation coverage, and test_block_sparse_end_step_precision_forwarded covers forwarding of both new knobs.

File: vllm_omni/__init__.py [hunk 7]
```diff
@@ -18,6 +21,21 @@
 # throw in patch if the versions differ.
 from .version import __version__, __version_tuple__  # isort:skip # noqa: F401
 
+# Early-register MindIE-SD custom ops (QBSA etc.) before CANN snapshots the
```
Comment (david6666666) [inline vllm_omni/__init__.py]: please don't modify here

**Review by david6666666 — CHANGES_REQUESTED** [review]

File: vllm_omni/__init__.py (hunk 7 above)
Comment (HAAZZZEEEE) [reply vllm_omni/__init__.py]: OK

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py [hunk 8]
```diff
@@ -50,6 +50,26 @@
...
 
+# Whether the installed mindiesd ``sparse_attention`` accepts ``precision=``.
+# Releases without it accept the kwarg through ``**kwargs`` but silently ignore
+# it, so a requested mix/fp8 mode would silently run the BF16 path. Cached
+# because forwards run per layer per denoise step.
+_MINDIESD_HAS_PRECISION: bool | None = None
+
+
+def _mindiesd_supports_precision() -> bool:
+    global _MINDIESD_HAS_PRECISION
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/rainfusion_attn.py]: why use global var, please change

**Review by david6666666 — DISMISSED** [review]
please add A/B test to show performance gain, and paste your video output comparasion

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py (hunk 8 above)
Comment (HAAZZZEEEE) [reply vllm_omni/diffusion/attention/backends/rainfusion_attn.py]: Fixed: replaced the module-level `_MINDIESD_HAS_PRECISION` + `global` memoization with `functools.cache` — the cache is now encapsulated in the function object instead of the module namespace.

Rationale: the probe has a single call site (the gate in `_forward_sparse_npu`) that only reads the return value, so there is no cross-function sharing need that would justify a `global`; but it runs per layer per denoise step, so caching is still required. `functools.cache` keeps the function pure and the namespace clean.

Behavior unchanged: first call probes, later calls hit the cache. Covered by `test_precision_non_bf16_requires_mindiesd_support` and `test_precision_non_bf16_passes_gate_when_supported` (both pass).

**Comment by HAAZZZEEEE** [comment]
> please add A/B test to show performance gain, and paste your video output comparasion
OK, already in test result

File: vllm_omni/diffusion/attention/backends/rainfusion_attn.py [hunk 9]
```diff
@@ -446,6 +483,7 @@ def _forward_sparse_npu(
             "inner_precise": _INNER_PRECISE,
             "block_size": _BLOCK_SIZE,
             "sparsity": self.rainfusion.sparsity,
+            "precision": self.rainfusion.precision,
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/rainfusion_attn.py:486]: Ran tests/e2e/accuracy/minimax_h3/test_minimax_h3_i2va_ref2va_similarity.py on 4x B300 with seed 0, 50 steps, USP4/HSDP4, and FLASH_ATTN: both I2VA and Ref2VA passed the SSIM/PSNR gates. The I2VA output is byte-identical to base (SSIM 0.972621, PSNR 35.074173 dB). Since this precision argument is only used by _forward_sparse_npu, no GPU accuracy regression was observed.

**Review by david6666666 — APPROVED** [review]
